unlock-app/src/components/interface/checkout/alpha/Checkout/CardPayment.tsx

```diff
@@ -106,82 +123,117 @@ export function CardPayment({ checkoutService, injectedProvider }: Props) {
   )
 }
 
-interface CardFormProps {
-  isSaving: boolean
-  setIsSaving(value: boolean): void
-  onSave(): void
+interface SetupProps {
+  onSubmit(): void
+  onSubmitted(intent?: SetupIntentResult): void
+  stripe: Promise<Stripe | null>
+}
+
+export function Setup({ onSubmit, stripe, onSubmitted }: SetupProps) {
+  const storageService = useStorageService()
+  const [clientSecret, setClientSecret] = useState('')
+  const walletService = useWalletService()
+  const { account, network } = useAuth()
+
+  const fetchSecret = useCallback(async () => {
+    await storageService.loginPrompt({
+      walletService,
+      address: account!,
+      chainId: network!,
+    })
+    const secret = await storageService.getSetupIntent()
+    setClientSecret(secret)
+  }, [storageService, account, network, walletService])
+
+  useEffect(() => {
+    if (!clientSecret) {
+      fetchSecret()
+    }
+  }, [fetchSecret, clientSecret])
+
+  if (!clientSecret) {
+    return null
+  }
+
+  return (
+    <Elements
+      stripe={stripe}
+      options={{
+        clientSecret,
+        appearance: {
+          theme: 'stripe',
+
+          variables: {
+            colorBackground: '#fff',
+          },
+        },
+      }}
+    >
+      <PaymentForm onSubmit={onSubmit} onSubmitted={onSubmitted} />
+    </Elements>
+  )
+}
+
+interface PaymentFormProps {
+  onSubmit(): void
+  onSubmitted(intent?: SetupIntentResult): void
 }
 
-function CardForm({ onSave, setIsSaving }: CardFormProps) {
+export function PaymentForm({ onSubmit, onSubmitted }: PaymentFormProps) {
+  const stripe = useStripe()
+  const elements = useElements()
   const {
     register,
     formState: { errors },
     handleSubmit,
-  } = useForm()
-  const stripe = useStripe()
-  const elements = useElements()
-  const { account } = useAuth()
-  const config = useConfig()
-  const walletService = useWalletService()
+  } = useForm<{
+    name: string
+  }>()
 
-  const onSubmit = async (data: FieldValues) => {
-    setIsSaving(true)
-    const cardElement = elements!.getElement(CardElement)
-    const result = await stripe!.createToken(cardElement!, {
-      address_country: data.address_country,
-      name: data.name,
-    })
-    if (result.token && account) {
-      deleteCardForAddress(config, walletService, account)
-      await saveCardsForAddress(config, walletService, account, result.token.id)
+  const onHandleSubmit = async ({ name }: Record<'name', string>) => {
+    if (!stripe || !elements) {
+      return
     }
-    onSave()
-  }
 
-  let errorMessage = ''
-  if (errors?.name?.message) {
-    errorMessage = errors.name.message as unknown as string
-  }
+    onSubmit()
+
+    const { error, setupIntent } = await stripe.confirmSetup({
+      elements,
+      redirect: 'if_required',
```

Comment:
Did you solve the redirection issue you were asking for?

Reply:
So re-direct isn't mandatory for card & google pay/apple pay. This won't end up redirecting unless we enable more payment option.

I tried a hack to solve redirect by storing the state in local storage and adding a flag to reload from it but I think it's unnecessary for now so instead of hack, we will come back when we store information about the flow in the routes properly to load them up without local storage.